snapcore/snapcraft
# repo: catch error updating the package cache

-----
This PR handles errors when updating the package cache, such as a hash mismatch or unresolved hostnames. The apt API doesn't distinguish these, so unfortunately we can't be very specific here but need to rely on the output we get before the error is raised.
If the error is a hashsum mismatch, the index will be cleared and Snapcraft will continue.

Fixes: #2029 #2030 
Fixes: [LP: #1747715](https://bugs.launchpad.net/snapcraft/+bug/1747715)

The following tests are added/ modified:

 - tests.unit.test_errors
 - To verify CacheUpdateFailedError
 - tests.unit.repo.test_deb.UbuntuTestCase.test_cache_update_failed
 - To verify that an `apt.cache.FetchFailedException` is handled as a proper error.
 - tests.unit.repo.test_deb.UbuntuTestCase.test_cache_hashsum_mismatch
 - To verify that a hashsum error will be caught and proceed after clearing the index.

I locally ran the tests:
 - `./runtests.sh tests/unit` with unrelated failures in [tests.unit.test_lifecycle.CoreSetupTestCase.test_core_setup_if_docker_env](https://bugs.launchpad.net/snapcraft/+bug/1752576), `tests.unit.test_mangling.TestClearExecstack.test_execstack_clears` and `tests.unit.test_elf`.
 - `./runtests.sh tests/integration` with one unrelated failure in [tests.integration.general.test_parser.TestParserWikis](https://bugs.launchpad.net/snapcraft/+bug/1752580)
 - `./runtests.sh static`: Everything passed

Manual test steps:
 - cd tests/integration/snaps/go-hello
 - Add a stage package:
   stages-packages: [appstream]
 - Modify `/etc/apt/sources.list` so that it can't update successfully, such as by adding an invalid URL.
 - snapcraft
 - Observe the error "Failed to update the package cache: Some files could not be downloaded. Check that the sources on your host are configured correctly.".
